packages/ai-jsx/src/core/conversation.tsx

```diff
@@ -0,0 +1,295 @@
+ *      <UserMessage>Hello!</UserMessage>
+ *    </ChatCompletion>
+ *
+ *    ==> 'Hello there!'
+ * ```
+ *
+ */
+export async function* Converse(
+  {
+    reply,
+    children,
+  }: {
+    reply: (messages: ConversationMessage[], fullConversation: ConversationMessage[]) => AI.Renderable;
+    children: AI.Node;
+  },
+  { render, memo }: AI.ComponentContext
+): AI.RenderableStream {
+  yield AI.AppendOnlyStream;
+
+  let fullConversation = [] as ConversationMessage[];
+  let next = memo(children);
+  while (true) {
+    const messages = await renderToConversation(next, render);
+    if (messages.length === 0) {
+      break;
+    }
+
+    fullConversation = fullConversation.concat(messages);
+    next = memo(reply(messages, fullConversation));
+    yield next;
+  }
+
+  return AI.AppendOnlyStream;
+}
+
+/**
+ * Allows the presentation of conversational components ({@link UserMessage} et al) to be altered.
+ *
+ * Also accepts an `onComplete` prop, which will be invoked once per render with the entire conversation.
+ *
+ * @example
+ * ```tsx
+ *     <ShowConversation present={(msg) => msg.type === "assistant" && <>Assistant: {msg.element}</>}>
+ *         <UserMessage>This is not visible.</UserMessage>
+ *         <Assistant>This is visible!</UserMessage>
+ *     </ShowConversation>
+ *
+ *     ==> 'Assistant: This is visible!'
+ * ```
+ */
+export async function* ShowConversation(
+  {
+    children: children,
+    present,
+    onComplete,
+  }: {
+    children: AI.Node;
+    present?: (message: ConversationMessage) => AI.Node;
+    onComplete?: (conversation: ConversationMessage[], render: AI.RenderContext['render']) => Promise<void> | void;
+  },
+  { render, isAppendOnlyRender, memo }: AI.ComponentContext
+): AI.RenderableStream {
+  // If we're in an append-only render, do the transformation in an append-only manner so as not to block.
+  if (isAppendOnlyRender) {
+    yield AI.AppendOnlyStream;
+  }
+
+  let lastFrame = [] as AI.PartiallyRendered[];
+  let conversationMessages = [] as ConversationMessage[];
+
+  function handleFrame(frame: AI.PartiallyRendered[]): AI.Node {
+    if (isAppendOnlyRender) {
+      const delta = toConversationMessages(frame.slice(lastFrame.length));
+      conversationMessages = conversationMessages.concat(delta);
+      lastFrame = frame;
+      return delta.map(present ?? ((m) => m.element));
+    }
+
+    conversationMessages = toConversationMessages(frame);
+    return toConversationMessages(frame).map(present ?? ((m) => m.element));
+  }
+
+  // Memoize before rendering so that the all the conversational components get memoized as well.
+  const finalFrame = yield* render(memo(children), {
+    map: handleFrame,
+    stop: isConversationalComponent,
+    appendOnly: isAppendOnlyRender,
+  });
+
+  // Prioritize rendering by yielding the final frame before running the `onComplete` handler.
+  yield handleFrame(finalFrame);
```

Comment:
I feel like we have this issue of wanting `map` to apply to every frame, not just the intermediate ones in multiple places.

Reply:
Yeah, these `map` scenarios are very awkward. I've delayed changing the API because my intuition is roughly "we shouldn't need to be doing this" but the better option hasn't really emerged yet so I should probably just fix it.